### DISCUSSION TITLE
Python Examples for API Calls now that Authorisation is sorted

### DISCUSSION THREAD

**Discussion by Pyromancer579** [timdorr/tesla-api#301] [Q&A]
With lots of help, I have managed to implement @timdorr 's work thanks to @Skimike02 and @Dvermagithub and it looks like the tokens now work for me.

But the code I had previously been using for API calls to control my Powerwall (switch its modes) when my conditions were met, doesn't seem to work,  I've put the code below and the response I get:

    256    async def Boost():
    257        client = TeslaApiClient(TES_ID, TES_PW)
    258        energy_sites = await client.list_energy_sites()
    259        assert(len(energy_sites)==1)
    260        N_MODE = await energy_sites[0].get_operating_mode()
    261        if N_MODE==C_MODE:
    262            await energy_sites[0].set_operating_mode('backup')
    263            N_MODE = await energy_sites[0].get_operating_mode()
    264        await client.close()

**File "/home/pi/ZappiPowerwall/QuaterHourCheck.py", line 258, in Boost
energy_sites = await client.list_energy_sites()
File "/home/pi/.local/lib/python3.7/site-packages/tesla_api/init.py", line 106, in list_energy_sites
return [_class(self, products['energy_site_id']) for products in await self.get('products')]
File "/home/pi/.local/lib/python3.7/site-packages/tesla_api/init.py", line 79, in get
await self.authenticate()
File "/home/pi/.local/lib/python3.7/site-packages/tesla_api/init.py", line 67, in authenticate
expiry_time = timedelta(seconds=self.token['expires_in'])
KeyError: 'expires_in'**

Can anyone point me in the direction of some example Python API calls for use with the new token system? so I can edit my code and make it work again.

**Comment by skimike02** [comment]
@Pyromancer579  Here are the only two API calls I am currently making with the new token system:
```python3
def GetProducts(access_token):
    url="https://owner-api.teslamotors.com/api/1/products"
    headers = {"Authorization": "Bearer "+access_token}
    payload={}
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
        return a
    else:
        a=json.loads(r.content.decode('utf-8'))
        response=a['response']
        return response
    
def GetSolarHistory(access_token,site_id,date):
    url=f"https://owner-api.teslamotors.com/api/1/energy_sites/{site_id}/calendar_history"
    headers = {"Authorization": "Bearer "+access_token}
    payload={}
    kind='power'
    period='day'
    end_date = date.replace(hour=7, minute=59, second=59)+datetime.timedelta(days=1)
    end_date=end_date.isoformat()
    params={'kind':kind,'period': period}
    params['end_date']=end_date+'Z'
    r = requests.get(url, data=json.dumps(payload), headers=headers, params=params)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
    else:
        a=json.loads(r.content.decode('utf-8'))
    return a['response']
```
All API calls work the same in the new system as the old system. You just have to pass a new access token. It looks like your old calls did something with the token expiration date. Maybe you were checking to see if the token needed refreshing each time? You could just delete those parts of the code. My implementation is to check once a day to refresh tokens, then when making API calls, just assume the tokens are good. Alternatively, you could try the access token, and only try to refresh if it doesn't work.

I would be curious what the structure of self.token in your code used to be, as it had a ['expires_in'] attribute previously. You will need to feed your API the same format now.

**Reply by Pyromancer579** [reply]
@skimike02  Thanks, that has helped a little:

Using your GetProducts routine with the 'return response' disabled and replaced with some print statements so I can see what is happening:

    def GetProducts(access_token):
        url="https://owner-api.teslamotors.com/api/1/products"
        headers = {"Authorization": "Bearer "+access_token}
        payload={}
        r = requests.get(url, data=json.dumps(payload), headers=headers)
        if r.status_code!=200:
            a="Error: http response "+str(r.status_code)+" "+r.reason
            #return a
            print ("a = ",a)
        else:
            a=json.loads(r.content.decode('utf-8'))
            response=a['response']
            #return response
            print (" ")
            print ("Response = ",response)
I get:
     
**Response =  [{'energy_site_id': ???????, 'resource_type': 'battery', 'site_name': 'My Address', 'id': 'Kind Of My Name + Some Numbers', 'gateway_id': '1152100-03-E--TG????????????', 'asset_site_id': 'af80c37d-972c-4d6b-bca4-779609c368a5', 'energy_left': 8368.473684210529, 'total_pack_energy': 13699, 'percentage_charged': 61.088208513107, 'battery_type': 'ac_powerwall', 'backup_capable': True, 'battery_power': 590, 'sync_grid_alert_enabled': True, 'breaker_alert_enabled': False, 'components': {'battery': True, 'battery_type': 'ac_powerwall', 'solar': True, 'grid': True, 'load_meter': True, 'market_type': 'residential'}}]**
 
This shows an 'Energy Site' presumably my Gateway 2 that interfaces my grid, solar and Powerwall (battery).  The code I was originally using changed my Powerwall (Gateway 2) Modes using 'energy_site' calls - so that makes sense to me.

According to [https://www.teslaapi.io/energy-sites/state-and-settings](https://www.teslaapi.io/energy-sites/state-and-settings), I need to use _**`https://owner-api.teslamotors.com/api/1/energy_sites/:site_id/status`**_ to take a look at an energy site.

I have tried each and every 'id' from the Response to GetProduct, with and without a colon infront, to replace **':site_id'** - for example
 
 ```
def Normal(access_token):
    url="https://owner-api.teslamotors.com/api/1/energy_sites/Kind Of My Name + Some Numbers/status"
    headers = {"Authorization": "Bearer "+access_token}
    payload={}
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    print (" ")
    print ("Normal Response = ",r)
    print ("Normal Payload = ",json.dumps(payload))
```
But all I get is 404 Site Not found: 
 
**Normal Response =  <Response [404]>
Normal Payload =  {}**

Any suggestions?

**Reply by skimike02** [reply]
I think something is missing in the documentation. This code works for "live_status", but not for "status." I got a 404 on status.

site_id should be 'energy_site_id'

```python3
def GetStatus(access_token,site_id):
    url=f"https://owner-api.teslamotors.com/api/1/energy_sites/{site_id}/live_status"
    headers = {"Authorization": "Bearer "+access_token}
    payload=[]
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
    else:
        a=json.loads(r.content.decode('utf-8'))
    return a['response']
```
I got this in return:
```
{'solar_power': 1390,
 'energy_left': 27841,
 'total_pack_energy': 27841,
 'percentage_charged': 100,
 'backup_capable': True,
 'battery_power': 0,
 'load_power': 317,
 'grid_status': 'Active',
 'grid_services_active': False,
 'grid_power': -1073,
 'grid_services_power': 0,
 'generator_power': 0,
 'storm_mode_active': False,
 'timestamp': '2021-02-08T14:55:29-08:00'}
```
also, your environment should have a way for you to look at variables without writing print statements. were you on VS code? https://visualstudiomagazine.com/articles/2019/04/24/vs-code-python-update.aspx

**Reply by Pyromancer579** [reply]
@skimike02  I'm a novice Python coder (just started tinkering last year), last coded in anger back in the 1990's when it was called 'Programming', so simple print statements make sense to me - I'll take a look at VS code thanks for the link.

I'll also try 'live_status' this evening - I have to attend to the day job first.

With regard to the ['expires_in'] attribute for tokens, that you were interested in - I can't really help, as this was all handled by 3rd Party routines that were all running invisibly to me, in my original implementation.  If you're really, really interested I'll try and dig back through my notes and point you to where I downloaded them from (I'm pretty sure it was somewhere on Github).

**Reply by Pyromancer579** [reply]
@skimike02  I'm coding and testing in 'Thonny Python IDE' on my Pi, so I don't think I can use VS code, I can see the benefit of it and if I was doing more than just this 'simple' application then I'd set it up on my PC.

Anyhow, I have tried 'live_status'and it works - Thanks.

I then worked my way through all the states and setting, commands, etc. that I could find on [https://www.teslaapi.io/energy-sites/state-and-settings](https://www.teslaapi.io/energy-sites/state-and-settings), for Energy Sites and Powerwalls.

To do this I first ran the Token Authentication that is now working, then your GetProducts routine (with my beloved print statements):
```
def GetProducts(access_token):
    url="https://owner-api.teslamotors.com/api/1/products"
    headers = {"Authorization": "Bearer "+access_token}
    payload={}
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
        #return a
        print ("a = ",a)
    else:
        a=json.loads(r.content.decode('utf-8'))
        response=a['response']
        #return response
        print (" ")
        print ("Product List = ",response) 
```
Which delivers: 

**Product List =  [{'energy_site_id': ???????, 'resource_type': 'battery', 'site_name': 'My Address', 'id': 'Kind Of My Name + Some Numbers', 'gateway_id': '1152100-03-E--TG????????????', 'asset_site_id': 'af80c37d-972c-4d6b-bca4-779609c368a5', 'energy_left': 8368.473684210529, 'total_pack_energy': 13699, 'percentage_charged': 61.088208513107, 'battery_type': 'ac_powerwall', 'backup_capable': True, 'battery_power': 590, 'sync_grid_alert_enabled': True, 'breaker_alert_enabled': False, 'components': {'battery': True, 'battery_type': 'ac_powerwall', 'solar': True, 'grid': True, 'load_meter': True, 'market_type': 'residential'}}]**

I then run my Normal routine (now tweaked to allow me to easily try the different variables by editing the value for 'test'):

```
def Normal(access_token):
    energy_site_id="???????"
    test="/live_status"
    testing=energy_site_id+test
    url="https://owner-api.teslamotors.com/api/1/energy_sites/"+testing
    headers = {"Authorization": "Bearer "+access_token}
    payload={}
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
        #return a
        print (" ")
        print ("a = ",a)
    else:
        a=json.loads(r.content.decode('utf-8'))
        response=a['response']
        #return response
        print (" ")
        print ("Test = ",test," Result = ",response)
    print (" ")
```
 Which Delivered:
 
**Test =  /live_status  Result =  {'solar_power': 1736, 'energy_left': 12168.473684210527, 'total_pack_energy': 13699, 'percentage_charged': 88.82745955332891, 'backup_capable': True, 'battery_power': -1060, 'load_power': 676, 'grid_status': 'Active', 'grid_services_active': False, 'grid_power': 0, 'grid_services_power': 0, 'generator_power': 0, 'storm_mode_active': False, 'timestamp': '2021-02-09T15:16:43Z'}**

I have tried the variables shown on the TESLA API pages linked above and only **'/live_status'** and **'/site_info'** work - the latter just delivering my **'gateway_id'** from the Product List.

All my other attempts have resulted in 404 Errors:

/status
/history
/backup
/site_name
/operation
/time_of_use_settings
/storm_mode
/powerhistory              - Powerwall call, but had a try anyway - still 404 Error.
/energyhistory             - Powerwall call, but had a try anyway - still 404 Error.

Can anyone point me toward the correct calls/commands?

I just need to establish my current Powerwall/Gateway Mode (**'backup'**, **'self_consumption'** or **'autonomous'**) and then get my Powerwall/Gateway to switch between these Modes.

**Reply by skimike02** [reply]
to get the status (backup/automated/self-powered), you need to use
```
url=f"https://owner-api.teslamotors.com/api/1/powerwalls/{battery_id}"
```
this code should get you there if you feed it an access token:
```python3
import requests
import json
import datetime

def GetProducts(access_token):
    url="https://owner-api.teslamotors.com/api/1/products"
    headers = {"Authorization": "Bearer "+access_token}
    payload={}
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
        return a
    else:
        a=json.loads(r.content.decode('utf-8'))
        response=a['response']
        return response
    
def GetBatteries(access_token):
    battery_ids=[]
    products=GetProducts(access_token)
    for product in products:
         if 'energy_site_id' in product:
             battery_ids.append(product['id'])
    return battery_ids
    
def GetStatus(access_token,battery_id):
    url=f"https://owner-api.teslamotors.com/api/1/powerwalls/{battery_id}"
    headers = {"Authorization": "Bearer "+access_token}
    payload=[]
    r = requests.get(url, data=json.dumps(payload), headers=headers)
    if r.status_code!=200:
        a="Error: http response "+str(r.status_code)+" "+r.reason
    else:
        a=json.loads(r.content.decode('utf-8'))
    return a['response']

for battery_id in GetBatteries(access_token):
    print(GetStatus(access_token,battery_id))
```

**Reply by skimike02** [reply]
Code below should let you set the status. I got stuck on this for quite a while before I realized you are supposed to send the payload as json. It returned a 200, but I didn't check to see if it actually changes mode, since I don't like changing my mode. Let me know if it works, and what the modes are. I think they are "backup","self_consumption" and "autonomous" but I performed no testing.

```python3
def SetStatus(access_token,site_id,mode):
    url=f"https://owner-api.teslamotors.com/api/1/energy_sites/{site_id}/operation"
    UA = "Mozilla/5.0 (Linux; Android 10; Pixel 3 Build/QQ2A.200305.002; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/85.0.4183.81 Mobile Safari/537.36"
    X_TESLA_USER_AGENT = "TeslaApp/3.10.9-433/adff2e065/android/10"
    headers = {
        "User-Agent": UA,
        "x-tesla-user-agent": X_TESLA_USER_AGENT,
        "X-Requested-With": "com.teslamotors.tesla",
        "Authorization": "Bearer "+access_token}
    payload={"default_real_mode": mode}
    r = requests.post(url, json=payload, headers=headers)
    return r
```

**Reply by Pyromancer579** [reply]
Thanks, I'll digest, give this a try and report back.

**Reply by Pyromancer579** [reply]
I've had a play and GetProducts and GetBatteries work.

GetStatus does not work - 404 Error.

SetStatus does work.

I'll keep playing.

**Reply by Pyromancer579** [reply]
@skimike02  I tried using SetStatus as a basis for a new GetStatus with a .get replacing the .post, plus some other tweeks, but no luck, still getting 404 errors.

It would be really nice to be able to read the Status (Mode) of my Gateway (Energy Site) before changing it - Any thoughts?

If not, I'll have to brute force it using SetStatus and assume it has worked, rather than checking first and afterwards to see the status and then taking action accordingly.

**Reply by skimike02** [reply]
GetStatus worked for me. I have solar plus powerwall, if that makes a difference. Make sure you are using the battery id, not the site id. Short of that, not sure why it wouldn't work. Maybe you could look at what the request is. Run the following (make sure to strip out your access token and other sensitive data from the last print statement if you post the results here):
```python3
battery_id=GetBatteries(access_token)[0]
url=f"https://owner-api.teslamotors.com/api/1/powerwalls/{battery_id}"
headers = {"Authorization": "Bearer "+access_token}
payload=[]
r = requests.get(url, data=json.dumps(payload), headers=headers)
print(r)
print(r.content)
print(r.request.__dict__)
```

**Reply by Pyromancer579** [reply]
@skimike02 I too have solar PV and a Powerwall, but I only get the information about an 'Energy Site' (presumably my Gateway2) on the Products List - it only mentions a Powerwall under 'components', but it does show the Powerwall charge state as part of the the data associated with the 'Energy Site':

**Product List = [{'energy_site_id': ???????, 'resource_type': 'battery', 'site_name': 'My Address', 'id': 'Kind Of My Name + Some Numbers', 'gateway_id': '1152100-03-E--TG????????????', 'asset_site_id': 'af80c37d-972c-4d6b-bca4-779609c368a5', 'energy_left': 8368.473684210529, 'total_pack_energy': 13699, 'percentage_charged': 61.088208513107, 'battery_type': 'ac_powerwall', 'backup_capable': True, 'battery_power': 590, 'sync_grid_alert_enabled': True, 'breaker_alert_enabled': False, 'components': {'battery': True, 'battery_type': 'ac_powerwall', 'solar': True, 'grid': True, 'load_meter': True, 'market_type': 'residential'}}]**

 I think it's just a quirk of the way my system was set up by the installer, that causes my Powerwall not to be listed separately on the Products List and this stops your GetStatus code working.

My 'brute force' implementation is now up and running, I'll test it for a few days and then publish it on here and elsewhere so others can benefit.

Thanks for your help, I would never have gotten my system working again without your much appreciated input.

**Reply by skimike02** [reply]
The product list above looks like mine. ``` id:'Kind Of My Name + Some Numbers'``` is the battery_id needed for the code above. You could try putting that into the code above to see what response you get.

**Reply by Pyromancer579** [reply]
@skimike02 Your code does work, it was just my limited knowledge of Python and shoddy implementation that messed it up - My Bad!

Thanks to your code I can now read the status of my Powerwall.

My Brute Force system is working well, now to finesses it to minimise unneeded calls.

Thanks again for your assistance.
